File: Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1] * (len(nums)) #fill res with 1 for length of input array nums
        
        prefix = 1
        for i in range(len(nums)): # in >> get the indexs of input array nums
            print("before", res[i], prefix)
            res[i] = prefix # set res[i] = prefix 
            print("after", res[i], prefix)
            prefix *= nums[i] # multiply prefix by input index
        postfix = 1 
        for i in range(len(nums) - 1, - 1, -1): # iterate from the back forwards
            res[i] *= postfix # calculate existing value res(prefix forward values) * values from back forward (starting from 1)
            postfix *= nums[i] # get back value (current value * next forward input array value)
        return res

File: Product of Array Except Self.js
//Time complexity = O(N)
//Space Complexity = O(N)

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const answer = [];

    let prefix = 1; // prefix = 1 offsets first iteration of multiplication calculation to itself
    //calculate total of i * i+1 forwards
    for (let i = 0; i < nums.length; i++) {
        answer[i] = prefix; // update answer to current multiplication calculation going forwards
        prefix *= nums[i] // update prefix = prefix * input[index]
    }
    console.log(answer)
    let postfix = 1;
    for (let i = nums.length - 1; i >= 0; i--) {
        answer[i] *= postfix //calculate total of prefix * nums.length - 1 - 1 (backwards)
        postfix *= nums[i] // update postfix = postfix * input[backwards index]
    }
    return answer
};
